src/background.js

```diff
@@ -469,33 +445,19 @@ const rollout = {
         }
       }
     });
-  },
 
-  async main() {
-    // Listen to the captive portal when it unlocks
-    browser.captivePortal.onStateChanged.addListener(rollout.onReady);
-
-    // If the captive portal is already unlocked or doesn't exist,
-    // run the measurement
-    let captiveState = await browser.captivePortal.getState();
-    log("Captive state:", captiveState);
-    if ((captiveState === "unlocked_portal") ||
-        (captiveState === "not_captive")) {
-      await rollout.onReady({state: captiveState});
-    }
-
-  },
+    browser.captivePortal.onConnectivityAvailable.addListener(async () => {
+      let skipHeuristicsCheck = await rollout.getSetting(DOH_SKIP_HEURISTICS_PREF, false);
+      let shouldRunHeuristics = await stateManager.shouldRunHeuristics();
 
-  async onReady(details) {
-    // Now that we're here, stop listening to the captive portal
-    browser.captivePortal.onStateChanged.removeListener(rollout.onReady);
+      if (shouldRunHeuristics && !skipHeuristicsCheck) {
+        rollout.runStartupHeuristics();
+      }
+    });
 
-    // Only proceed if we're not behind a captive portal
-    if ((details.state !== "unlocked_portal") &&
-        (details.state !== "not_captive")) {
-      return;
-    }
+  },
 
+  async runStartupHeuristics() {
```

Comment:
Please move the `shouldRunHeuristics` and `skipHeuristicsCheck` logic into this function, it's duplicated in every call site.

Reply:
Done.